Correct path comparisons for fs view

psakievich opened the pull request:
I'm working through why the binary path comparisons are not correct when using filesystem views.
Pinging @becker33 and @nhanford

Commit:
Initial debug of copy views

psakievich commented on lib/spack/spack/relocate.py:
```diff
@@ -497,6 +497,7 @@ def _replace_prefix_bin(filename, byte_prefixes):
             length_compatible = len(new_bytes) <= len(orig_bytes)
             if not length_compatible:
+                print('Failing filename', filename)
                 raise BinaryTextReplaceError(orig_bytes, new_bytes)
```
currently trying to figure out how this occurred:
```
Failing filename /projects/wind/spack-manager/views/snapshots/._20210910/pxq6nab47hwwoiqbrc3j5ktotjuddytg/lib/libnetcdf.so.18.0.0
==> Error: New path longer than old path: binary text replacement not possible.
The new path b'/projects/wind/spack-manager/views/snapshots/._20210910/pxq6nab47hwwoiqbrc3j5ktotjuddytg' is longer than the old path b'/apps/mpi/4.0.3-ucx'.
```

psakievich replied:
Ah figured it out.  the traversal of the spec is also capturing externals like `/usr` and adds them to the list of `prefix_to_projection`. I don't think this traversal is necessary. for copy we just want to move binary from location `a` to `b` we don't need to check any other locations for the same binary right?

Commit:
Remove traversal of spec dependencies for prefix checks

Commit:
Get correct permissions

becker33 commented on lib/spack/spack/filesystem_view.py:
```diff
@@ -73,16 +73,13 @@ def view_copy(src, dst, view, spec=None):
-            spec.prefix: view.get_projection_for_spec(spec),
-            spack.paths.spack_root: view._root})
-
-        for dep in spec.traverse():
-            prefix_to_projection[dep.prefix] = \
-                view.get_projection_for_spec(dep)
```
I think we still need the dependency prefixes in the translation dict

psakievich replied:
Okay I've updated this to be the same, just excluding externals from the `prefix_to_projection`

Commit:
add back depedency checks

This time exclude externals from the check.
External paths are very likely shorter than the
original spack install, but they also don't need to change
for a copy operation on the same machine/env

becker33 commented on lib/spack/spack/filesystem_view.py:
```diff
@@ -96,6 +99,7 @@ def view_copy(src, dst, view, spec=None):
                 files=[dst],
                 prefixes=prefix_to_projection
             )
+        fp.set_permissions_by_spec(dst, spec)
```
Is this still necessary after switching to `copy2` above?

psakievich replied:
I haven't tested it. I don't think so but I do want to keep groups so something will have to be added in its place

psakievich replied:
updated this.

Commit:
Start a unit test

Commit:
Set group id though os

psakievich commented:
IDK what to do about adding a test for this to test the failures I found. I don't think we want to create/install an actual binary to replace paths and the actual path check is pretty low level.  I looked at the unit tests for `relocate` and that is exactly what is done.  I'm going to delete the test I added and mark this ready for review.  The feature itself is still tested, just not the path length checks that caused the failure that prompted this PR.

Commit:
Revert "Start a unit test"

This reverts commit 0ba85d25631101e89937f91ea1ed8e3dea5f1b82.

Commit:
Remove stale import

scottwittenburg commented:
@spackbot run pipeline

psakievich commented:
@spackbot re-run pipeline

becker33 commented on lib/spack/spack/filesystem_view.py:
```diff
@@ -96,6 +97,8 @@ def view_copy(src, dst, view, spec=None):
                 prefixes=prefix_to_projection
             )
+        stat = os.stat(src)
+        os.chown(dst, stat.st_uid, stat.st_gid)
```
Can you put a try/except block around this? This will fail without root permissions if you are not the original owner of the file or don't belong to the group it originally belonged to. I think we should be best-effort about this instead.

psakievich replied:
Done

becker33 reviewed (changes requested):
Mostly LGTM. One small request.

Commit:
Add try/except around permissions

psakievich committed:
Change debug info

psakievich commented:
@becker33 ping for merge
